File: job-agent-os/frontend/src/lib/api/jobs.ts
/** Jobs API */

import { api } from "./client";
import type { PaginatedData } from "@/types/api";
import type { JobFilters, JobResponse, ManualJobCreate } from "@/types/job";

export function listJobs(
  params: JobFilters & { page?: number; page_size?: number } = {}
) {
  const query = new URLSearchParams();
  if (params.platform) query.set("platform", params.platform);
  if (params.location) query.set("location", params.location);
  if (params.keyword) query.set("keyword", params.keyword);
  if (params.status) query.set("status", params.status);
  if (params.page) query.set("page", String(params.page));
  if (params.page_size) query.set("page_size", String(params.page_size));
  const qs = query.toString();
  return api.get<PaginatedData<JobResponse>>(`/v1/jobs${qs ? `?${qs}` : ""}`);
}

export function getJob(jobId: string) {
  return api.get<JobResponse>(`/v1/jobs/${jobId}`);
}

export function createManualJob(data: ManualJobCreate) {
  return api.post<JobResponse>("/v1/jobs/manual", data);
}

export function triggerJobSearch(data: {
  query_text?: string;
  platforms?: string[];
}) {
  return api.post<{ session_id: string }>("/v1/jobs/search", data);
}

File: job-agent-os/src/job_agent_os/api/v1/jobs.py
"""Jobs endpoints."""

from uuid import UUID

from fastapi import APIRouter, Query, status

from job_agent_os.api.deps import CurrentUser, DBSession, Pagination
from job_agent_os.api.response import paginated_response, success_response
from job_agent_os.schemas.job import (
    JobResponse,
    JobSearchRequest,
    ManualJobCreate,
)
from job_agent_os.services.job_service import JobService

router = APIRouter()


@router.post("/search", status_code=status.HTTP_202_ACCEPTED)
async def trigger_job_search(
    request: JobSearchRequest,
    db: DBSession,
    user: CurrentUser,
) -> dict:
    """Trigger an async job search. Returns session_id for polling."""
    service = JobService(db)
    structured = request.structured_query.model_dump() if request.structured_query else None
    session_id = await service.trigger_search(
        user=user,
        query_text=request.query_text,
        structured_query=structured,
        platforms=request.platforms,
    )
    return success_response(
        data={"session_id": str(session_id)},
        message="Search task accepted",
    )


@router.get("")
async def list_jobs(
    db: DBSession,
    user: CurrentUser,
    pagination: Pagination,
    platform: str | None = Query(default=None),
    location: str | None = Query(default=None),
    keyword: str | None = Query(default=None),
    job_status: str | None = Query(default=None, alias="status"),
) -> dict:
    """List jobs with pagination and filters."""
    service = JobService(db)
    jobs, total = await service.list_jobs(
        pagination=pagination,
        platform=platform,
        location=location,
        keyword=keyword,
        status=job_status,
    )
    items = [JobResponse.model_validate(j).model_dump(mode="json") for j in jobs]
    return paginated_response(
        items=items,
        total=total,
        page=pagination.page,
        page_size=pagination.page_size,
    )


@router.get("/{job_id}")
async def get_job(
    job_id: UUID,
    db: DBSession,
    user: CurrentUser,
) -> dict:
    """Get job details by ID."""
    service = JobService(db)
    job = await service.get_job(job_id)
    return success_response(data=JobResponse.model_validate(job).model_dump(mode="json"))


@router.post("/manual", status_code=status.HTTP_201_CREATED)
async def create_manual_job(
    request: ManualJobCreate,
    db: DBSession,
    user: CurrentUser,
) -> dict:
    """Manually create a job entry."""
    service = JobService(db)
    job = await service.create_manual_job(request)
    return success_response(data=JobResponse.model_validate(job).model_dump(mode="json"))

